### PR TITLE
Put all files into a common prefix in DropBox

### PR DESCRIPTION
**Pull request by Ragnis**

### CODE REVIEW

**Comment by skddc** [comment]
Welcome back, @Ragnis! :)

Looks like there are two failing Dropbox tests left to fix. Do you need help with it or wanna solve it on your own?

**Comment by skddc** [comment]
Any news on this one?

**Comment by Ragnis** [comment]
I haven't been able to get the tests running locally in order to fix them.

File: src/dropbox.js [hunk 1]
```diff
@@ -41,7 +41,28 @@
   var hasLocalStorage;
   var AUTH_URL = 'https://www.dropbox.com/1/oauth2/authorize';
   var SETTINGS_KEY = 'remotestorage:dropbox';
-  var cleanPath = RS.WireClient.cleanPath;
+  var PATH_PREFIX = '/remotestorage';
+
+  /**
+   * Function: mapPath(path)
```
Comment (galfert) [inline src/dropbox.js]: The documentation mentions a different function name than what's actually used (`mapPath` vs. `getDropboxPath`).

Comment (skddc) [inline src/dropbox.js]: Why not just fix it real quick?

**Comment by galfert** [comment]
Testing in [Webmarks](https://github.com/skddc/webmarks), this doesn't work for me. When I create a new Bookmark, I see a `PUT` request that responds with a 200. I can see the folders (`remotestorage/bookmarks/archive`) have been created in my Dropbox folder, but there are no files in there. There seems to be a `DELETE` request shortly after the `PUT` that deletes the file again.

**Comment by galfert** [comment]
I think I found the culprit. After the `PUT` it detects a conflict, because the path of the response doesn't correspond with the actual path here: https://github.com/remotestorage/remotestorage.js/blob/feature/796-dropbox-common-base-path/src/dropbox.js#L796.

**Comment by galfert** [comment]
I fixed the comparison, but didn't have time yet to also write a test for it.

`PUT` requests work for me now.

But when clicking on the Dropbox icon in the widget, I see this error in the console:

> remotestorage.js:13440 Uncaught TypeError: rs.sync.bind is not a function

The error happens here: https://github.com/remotestorage/remotestorage.js/blob/feature/796-dropbox-common-base-path/src/dropbox.js#L854

**Comment by galfert** [comment]
> I fixed the comparison, but didn't have time yet to also write a test for it.

Just saw that there are actually failing tests now. So it seems that there were already tests for this, but they were just not adjusted for the new paths yet.

**Comment by galfert** [comment]
Fixed those tests.

**Comment by galfert** [comment]
FYI: I found a general problem with the Dropxbox (and also GoogleDrive) initialization while working on #940.
When loading the app for the first time, the app sets the API keys only after the remoteStorage initialization has already happened. But the Dropbox backend only gets initialized, when there is an API key set.

So the widget shows the Dropbox icon, but the class hasn't been initialized, which results in an exception when clicking the dropbox icon.

When loading the app for the second time and the API key is loaded from localStorage before the Dropbox initialization, the Dropbox backend object will be created properly.

**Comment by galfert** [comment]
I created a new issue for the backend initialization bug at #941, it's not really related to this PR.

**Comment by skddc** [comment]
Are you saying it's ready for review/merge?

**Comment by galfert** [comment]
Ok, I fixed the remaining issue where Dropbox wouldn't hook itself up properly into the sync cycle.

I think this PR is ready for review/merge now.

**Comment by skddc** [comment]
Here's a deployed version of myfavoritedrinks for easy testing:

https://mfd-dropbox.5apps.com/

I'll leave that online. Ping me if you want deploy access on 5apps for that app.

**Comment by skddc** [comment]
LGTM and tested with the above app. I ran into some unrelated issues, but will test some more and create seperate GH issues if needed.

**Comment by skddc** [comment]
@Ragnis wanna have another look at what @galfert added there and try it out yourself?

**Comment by galfert** [comment]
Weird. Just noticed that Travis had 2 failing tests (which worked for me locally). I restarted the Travis build and now they are green `¯\_(ツ)_/¯`

File: src/dropbox.js [hunk 2]
```diff
@@ -811,7 +837,7 @@
           return Promise.reject(new Error("Cannot delete '" + path + "': too many files involved"));
         }
 
-        if (resp.status === 200) {
+        if (resp.status === 200 || resp.status === 404) {
```
Comment (gregkare) [inline src/dropbox.js]: Why is this now checking for 404 as well as 200? I think there should be a comment about this because it is not clear to me

**Comment by gregkare** [comment]
LGTM, I also tested this and only had unrelated issues (the same that @skddc encountered) and created an issue for it: #943

I added a question

I think it would be a good idea not to hard-code the API URI everywhere (https://api.dropbox.com/1/) Should I create a separate issue for it?

**Comment by galfert** [comment]
> I think it would be a good idea not to hard-code the API URI everywhere (https://api.dropbox.com/1/) Should I create a separate issue for it?

I created an issue for it at #944.

**Comment by gregkare** [comment]
@galfert I think you missed my question (https://github.com/remotestorage/remotestorage.js/pull/928#discussion_r71178250). Does anyone else know about this change?

File: src/dropbox.js (hunk 2 above)
Comment (skddc) [inline src/dropbox.js]: It's a delete request, so if it tries to delete something that doesn't exist, that would result in a 404 I think.

Comment (gregkare) [inline src/dropbox.js]: That makes sense, thanks

**Comment by gregkare** [comment]
+1

**Pull request merged**